pep-0635.rst

```diff
@@ -0,0 +1,689 @@
+
+Early ideas to induce a hierarchy on numbers so that ``case 1.0`` would
+match both the integer ``1`` and the floating point number ``1.0``, whereas
+``case 1:`` would only match the integer ``1`` very eventually dropped in
+favour of the simpler and consistent rule based on equality.
+
+Recall that literal patterns are *not* expressions, but directly denote a
+specific value or object.  From a syntactical point of view, we have to
+ensure that negative and complex numbers can equally be used as patterns,
+although they are not atomic literal values (i.e. ``-3+4j`` is syntactically
+an expression of the form ``BinOp(UnaryOp('-', 3), '+', 4j)``).
+Interpolated *f*-strings, on the other hand, are not literal values, despite
+their appearance and can therefore not be used as literal patterns.
+
+Literal patterns not only occur as patterns in their own right, but also
+as keys in *mapping patterns*.
+
+
+.. _constant_value_pattern:
+
+Constant Value Patterns
+~~~~~~~~~~~~~~~~~~~~~~~
+
+It is good programming style to use named constants for parametric values or
+to clarify the meaning of particular values.  Clearly, it would be desirable
+to also write ``case (HttpStatus.OK, body):`` rather than
+``case (200, body):``, say.  The main issue that arises here is how to
+discern capture patterns (variables) and constant value patterns.  The
+general discussion surrounding this issue has brought forward a plethora of
+options, which we cannot all fully list here.
+
+Strictly speaking, constant value patterns are not really necessary, but
+could be implemented using guards, i.e.
+``case (status, body) if status == HttpStatus.OK:``.  Nonetheless, the
+convenience of constant value patterns is unquestioned and obvious.
+
+The observation that constants tend to be written in uppercase letters or
+collected in enumeration-like namespaces suggests possible rules to discern
+constants syntactically.  However, the idea of using upper vs. lower case as
+a marker has been met with scepticism since there is no similar precedence
+in core Python (although it is common in other languages).  We therefore only
+adopted the rule that any dotted name (i.e. attribute access) is to be
+interpreted as a constant value pattern as exemplified by ``HttpStatus.OK``
+above.  This excludes, in particular, local variables from acting as
+constants.
+
+Global variables can only be directly used as constent when defined in other
+modules, although there are work arounds to access the current module as a
+namespace as well.  A proposed rule to use a leading dot (e.g.
+``.CONSTANT``) for that purpose was critisised because it was felt that the
+dot would not be a visible-enough marker for that purpose.  Partly inspired
+by use cases in other programming languages, a number of different
+markers/sigils was proposed (such as ``^CONSTANT``, ``$CONSTANT``,
+``==CONSTANT``, etc.), although there was no obvious or natural choice.
```

Comment:
Probably add `CONSTANT?` and definitely CONSTANT enclosed in backticks (which is hard to spell in markdown so I'd just say it in words).

Reply:
OK.